src/app/services/theme/theme.service.ts

```diff
@@ -0,0 +1,202 @@
+import { DOCUMENT } from "@angular/common";
+import { Inject, Injectable } from "@angular/core";
+import { IS_SERVER_PLATFORM } from "src/app/app.helper";
+
+export const bawThemes = [
+  "highlight",
+  "primary",
+  "secondary",
+  "success",
+  "info",
+  "warning",
+  "danger",
+  "light",
+  "dark",
+] as const;
+
+export const bawThemeVariants = [
+  "",
+  "-lighter",
+  "-lightest",
+  "-darker",
+  "-darkest",
+] as const;
+
+export type BawTheme = typeof bawThemes[number];
+export type BawThemeVariant = typeof bawThemeVariants[number];
+
+export interface RGB {
+  red: number;
+  green: number;
+  blue: number;
+}
+
+export interface HSL {
+  hue: number;
+  saturation: number;
+  lightness: number;
+}
+
+@Injectable({
+  providedIn: "root",
+})
+export class ThemeService {
+  public constructor(
+    @Inject(DOCUMENT) private document: Document,
+    @Inject(IS_SERVER_PLATFORM) private isServer: boolean
+  ) {}
+
+  /**
+   * Set the base colour for a theme
+   *
+   * @param theme Theme to modify
+   * @param color New base colour for theme
+   */
+  public setTheme(theme: BawTheme, color: string): void {
+    if (!this.validateHexRgb(color)) {
+      console.warn(`Invalid theme color given for ${theme} detected: ${color}`);
+      return;
+    }
+
+    const hsl = this.rgb2hsl(color);
+    this.updateTheme(
+      theme,
+      `
+        :root {
+          --baw-${theme}-hue: ${hsl.hue}deg;
+          --baw-${theme}-saturation: ${hsl.saturation}%;
+          --baw-${theme}-lightness: ${hsl.lightness}%;
+        }
+      `
+    );
+  }
+
+  /**
+   * Reset any modifications to the base colour of a theme
+   * ! This includes changes set by the environment
+   *
+   * @param theme Theme to reset
+   */
+  public resetTheme(theme: BawTheme): void {
+    const style = this.document.getElementById(this.getThemeId(theme));
+    if (style) {
+      this.document.head.removeChild(style);
+    }
+  }
+
+  /**
+   * Update a theme by either inserting, or modifying, a style in the header
+   *
+   * @param theme Theme to update
+   * @param styles CSS styles to apply
+   */
+  private updateTheme(theme: BawTheme, styles: string): void {
```

Comment:
Annoying that a literal style element is needed. Under what condition would it be missing?